crates/ecs/src/lib.rs

```diff
@@ -503,74 +499,348 @@ pub struct World {
+            .expect(
+            "Archetype should exist since it was the requirement for entering this matching case",
+        );
+
+        target_archetype
+            .entity_to_component_index
+            .insert(entity, target_archetype.entity_to_component_index.len());
+
+        target_archetype.entity_order.push(entity);
+        self.entity_to_archetype_index
+            .insert(entity, target_archetype_idx);
+
+        Ok(())
+    }
+
+    fn move_entity_components_to_new_archetype(
+        &mut self,
+        entity: Entity,
+        components_to_move: Vec<TypeId>,
+    ) -> WorldResult<ArchetypeIndex> {
+        let mut target_archetype = Archetype::default();
+        let target_archetype_idx = self.archetypes.len();
+        let source_archetype_idx = *self
+            .entity_to_archetype_index
+            .get(&entity)
+            .ok_or(WorldError::EntityDoesNotExist(entity))?;
+        let source_archetype = self
+            .archetypes
+            .get(source_archetype_idx)
+            .ok_or(WorldError::ArchetypeDoesNotExist(source_archetype_idx))?;
+        let source_component_vec_idx = *source_archetype
+            .entity_to_component_index
+            .get(&entity)
+            .ok_or(WorldError::EntityDoesNotExist(entity))?;
+
+        for type_id in components_to_move {
+            let source_component_vec = source_archetype
+                .component_typeid_to_component_vec
+                .get(&type_id)
+                .expect(
+                    "Type that tried to be fetched should exist
+                            in archetype since types are fetched from this archetype originally.",
+                );
+            source_component_vec
+                .move_element_to_new(source_component_vec_idx, &mut target_archetype)
+                .map_err(WorldError::CouldNotMoveComponent)?;
 
-        if !big_archetype.contains_component_type::<ComponentType>() {
-            let component_typeid = TypeId::of::<ComponentType>();
-            self.stored_types.push(component_typeid);
-            let archetype_index = 0;
             self.component_typeid_to_archetype_indices
-                .entry(component_typeid)
-                .or_default()
-                .insert(archetype_index);
+                .get_mut(&type_id)
+                .expect("Type ID should exist for previously existing types")
+                .insert(target_archetype_idx);
+        }
+
+        target_archetype.entity_order.push(entity);
+        target_archetype.entity_to_component_index.insert(entity, 0);
+        self.archetypes.push(target_archetype);
+        self.entity_to_archetype_index
+            .insert(entity, target_archetype_idx);
+
+        Ok(target_archetype_idx)
+    }
+
+    fn update_source_archetype_after_entity_move(
+        &mut self,
+        entity: Entity,
+        source_archetype_index: ArchetypeIndex,
+        source_component_vec_idx: ComponentIndex,
```

Comment:
You don't need to pass `source_component_vec_idx` as you already pass `Entity`, the calling code does not use `source_component_vec_idx`, so it is redundant to pass it. Instead move the code that gets this variable inside this function